gh-89263: Add typing.get_overloads (#31716)
python/cpython: Doc/library/functools.rst

   .. versionchanged:: 3.7
      The :func:`register` attribute now supports using type annotations.

   .. versionchanged:: 3.11

>       Implementation functions are now registered using :func:`register_variant`.

Reviewer: This feels a little opaque to me (same for the notice on line 592). Maybe we should wait until we've taught `inspect.signature()` to somehow use `get_variants` to give better results for multi-dispatch functions before we add these `.. versionchanged` notices? I feel like the benefit to the end user will be easier to explain at that point.

Author: I made it reference `get_variants` now, which is a bit more immediately useful to users.

   .. versionadded:: 3.8

   .. versionchanged:: 3.11

>       Implementation functions are now registered using :func:`register_variant`.

Author: ```suggestion
      Implementation functions can now be retrieved using :func:`get_variants`.
```

   would have been ``'wrapper'``, and the docstring of the original :func:`example`
   would have been lost.

.. function:: get_variants(func)

   Return all registered function variants for this function. Function variants are
   objects that represent some subset of the functionality of a function, for
   example overloads decorated with :func:`typing.overload` or :func:`singledispatch`
   implementation functions.

   Variants are registered by calling :func:`register_variant`.

   .. versionadded:: 3.11

.. function:: register_variant(func, variant)

>    Register *variant* for function *func* that can later be retrieved using

Reviewer 2: What are the types for `func` and `variant`? Maybe we need an example of how these three new functions work together?

   :func:`get_variants`.

   .. versionadded:: 3.11

.. function:: clear_variants(func=None)

>    Clear all registered variants for the given *func*. If *func* is None, clear

Reviewer 2: Describe the type of `func`.

   would have been ``'wrapper'``, and the docstring of the original :func:`example`
   would have been lost.

.. function:: get_variants(key)

   Return all registered function variants for this key. Function variants are

>    objects that represent some subset of the functionality of a function, for

Author: Not too happy with this wording. @AlexWaygood do you have suggestions?

Reviewer 2: At the very least describe the return type better. Is it a list? What are the items?

   example overloads decorated with :func:`typing.overload` or :func:`singledispatch`
   implementation functions.

   Variants are registered by calling :func:`register_variant`.
   The *key* argument is a string that uniquely identifies the function and its
   variants. It should be the result of a call to :func:`get_key_for_callable`.

   .. versionadded:: 3.11

.. function:: register_variant(key, variant)

   Register a function variant that can later be retrieved using
   :func:`get_variants`. The key should be the result of a call to
   :func:`get_key_for_callable`.

   .. versionadded:: 3.11

.. function:: clear_variants(key=None)

   Clear all registered variants with the given *key*. If *key* is None, clear
   all variants.

   .. versionadded:: 3.11

> .. function:: get_key_for_callable(func)
> 
>    Return a string key that can be used with :func:`get_variants` and
>    :func:`register_variant`. *func* must be a :class:`function`,
>    :class:`classmethod`, :class:`staticmethod`, or similar callable.
>    If no key can be computed, the function returns None.
> 
>    .. versionadded:: 3.11
> 

Reviewer: ```suggestion
```

Should delete the docs entry for this function now that it's been made private
